### PR TITLE
rgw/sfs: multipart

### PR DESCRIPTION
**Pull request by jecluis**
This patch set introduces proper multipart uploads in sfs.

~Right now though, we are not supporting aborting a multipart upload, nor aborting all multipart uploads in a bucket~. We are also not persisting on-going multipart state; this means that stopping the gateway will lead to losing all multipart information.

Tests can be found in https://github.com/aquarist-labs/s3gw-tools/pull/173 .

Signed-off-by: Joao Eduardo Luis \<joao@suse.com>

### CODE REVIEW

File: src/rgw/store/sfs/multipart.cc [hunk 1]
```diff
@@ -134,7 +118,143 @@ int SFSMultipartUpload::complete(
...
+    char *buf = (char *) std::malloc(block_size);
+    while (read_len < part->len) {
+      std::memset(buf, 0, block_size);
+      uint64_t to_read = std::min((part->len - read_len), block_size);
+      part_in.seekg(read_len);
+      if (!part_in.read(buf, to_read)) {
+        lsfs_dout(dpp, 0) << "error reading part " << parts_it->first
+                          << " from " << partpath
+                          << ", bytes: " << to_read << dendl;
+        return -ERR_INVALID_PART;
```
Comment (giubacc) [inline src/rgw/store/sfs/multipart.cc:198]: does `buf` leak here?

File: src/rgw/store/sfs/multipart.cc [hunk 2]
```diff
@@ -134,7 +118,143 @@ int SFSMultipartUpload::complete(
...
+      }
+
+      auto cur_offset = out.tellp();
+      out.seekp(0);
+      if (!out.write(buf, to_read)) {
+        lsfs_dout(dpp, 0) << "error writing part " << parts_it->first
+                          << " to " << outpath
+                          << ", bytes: " << to_read
+                          << ", offset: " << cur_offset << dendl;
+        return -ERR_INVALID_PART;
```
Comment (giubacc) [inline src/rgw/store/sfs/multipart.cc:208]: does `buf` leak here?

File: src/rgw/store/sfs/multipart.cc [hunk 3]
```diff
@@ -134,7 +118,143 @@ int SFSMultipartUpload::complete(
...
+                       << ", offset: " << ofs
+                       << dendl;
+    part_in.close();
+  }
+
+  out.flush();
+  out.close();
+
+  char final_etag[CEPH_CRYPTO_MD5_DIGESTSIZE];
+  char final_etag_str[(CEPH_CRYPTO_MD5_DIGESTSIZE * 2) + 16];
```
Comment (giubacc) [inline src/rgw/store/sfs/multipart.cc:226]: what is 16?
I'm not a big fan of magic numbers :)

File: src/rgw/store/sfs/multipart.h [hunk 4]
```diff
@@ -192,13 +182,25 @@ class SFSMultipartUpload : public MultipartUpload {
...
+
+  virtual int unlock() override {
+    return 0;
+  }
+
+};
+
 } // ns rgw::sal
 
 #endif // RGW_STORE_SFS_MULTIPART_H
```
Comment (giubacc) [inline src/rgw/store/sfs/multipart.h:206]: missing newline at the end of the file

File: src/rgw/store/sfs/multipart.cc (hunk 1 above)
Comment (jecluis) [reply src/rgw/store/sfs/multipart.cc:198]: good catch

File: src/rgw/store/sfs/multipart.cc (hunk 2 above)
Comment (jecluis) [reply src/rgw/store/sfs/multipart.cc:208]: good catch

File: src/rgw/store/sfs/multipart.cc (hunk 3 above)
Comment (jecluis) [reply src/rgw/store/sfs/multipart.cc:226]: this code was mostly lifted from dbstore's implementation, and at the time it felt it made sense. The 16 would seem to be meant to hold the `-%lld` below.
The more I think about it, the less sense it makes though. Yes, it is meant to hold `-%lld`, but that should either be 21 bytes (`-` plus 19 from the `LLONG_MAX` plus `\0`), or it should be 7 (`-` plus the number of characters in `10000` -- the maximum number of multipart parts --, plus `\0`).

I'll ensure we use the latter limits, and assert if something is weirdly wrong.

File: src/rgw/store/sfs/CHANGELOG.md [hunk 5]
```diff
@@ -15,10 +15,21 @@ and this project adheres to
...
 - Show delete markers when listing object versions.
 
+### Notes
+
+- Multipart uploads are currently tracked solely in memory. Should the gateway
+  be stopped, on-going multipart uploads will be lost.
+
+- Aborting on-going multipart uploads is currently not supported.
+
+- Listing multipart uploads does not account for prefix or delimiters.
```
Comment (giubacc) [inline src/rgw/store/sfs/CHANGELOG.md:29]: I would prefer we keep the existing convention where we are not inserting blank lines between list elements.
Not a real preference, just to respect what is the current convention.

File: src/rgw/store/sfs/types.h [hunk 6]
```diff
@@ -167,6 +341,55 @@ class Bucket {
...
+  std::string gen_multipart_upload_id() {
+    auto now = ceph::real_clock::now();
+    return ceph::to_iso_8601_no_separators(
+      now, ceph::iso_8601_format::YMDhmsn
+    );
+  }
+
+  std::map<std::string, MultipartUploadRef> get_multiparts() {
+    std::lock_guard l(multipart_map_lock);
+    // this should be doing a copy
```
Comment (giubacc) [inline src/rgw/store/sfs/types.h:410]: it is, I think we can get rid of this comment

**Review by giubacc — CHANGES_REQUESTED** [review]
just minor things, for the rest, ok.

**Comment by jecluis** [comment]
@0xavi0 @giubacc this is good to be reviewed now. Tests are available through https://github.com/aquarist-labs/s3gw-tools/pull/173 .

File: src/rgw/store/sfs/types.h (hunk 6 above)
Comment (jecluis) [reply src/rgw/store/sfs/types.h:410]: we can, but I left it there for a reason: to be clear we are doing a copy of the map, should someone less familiar with the code (or with c++) reads that and thinks we are returning the same map.

File: src/rgw/store/sfs/sqlite/objects/object_definitions.h [hunk 7]
```diff
@@ -47,4 +47,20 @@ struct DBOPObjectInfo {
...
+};
+
+struct DBOPMultipartObject {
+  uuid_d uuid;
+  std::string bucket_name;
+  std::string name;
+  std::string upload_id;
+  int state;
+};
+
```
Comment (0xavi0) [inline src/rgw/store/sfs/sqlite/objects/object_definitions.h:65]: I think this is not used, at least in this actual PR.
I guess it's here for making parts persist over crashes in the future?

**Review by 0xavi0 — COMMENTED** [review]
I need more time to fully review this but so far it looks good.

I've been testing and seems to be working fine.

One thing:
I see etags for parts are the same.
```
$ aws --endpoint=http://127.0.0.1:7480 s3api  list-parts --bucket  test --key test-large --upload-id 20220922T124511.426602599Z
{
    "Parts": [
        {
            "PartNumber": 1,
            "LastModified": "2022-09-22T12:47:16.450Z",
            "ETag": "\"e1e8d0676ebe133a7c9ce47c75011c1e\"",
            "Size": 126
        },
        {
            "PartNumber": 2,
            "LastModified": "2022-09-22T13:34:22.653Z",
            "ETag": "\"e1e8d0676ebe133a7c9ce47c75011c1e\"",
            "Size": 126
        }
    ],
    "ChecksumAlgorithm": null,
    "Initiator": null,
    "Owner": {
        "DisplayName": "",
        "ID": ""
    },
    "StorageClass": "STANDARD"
}
```

I also get the same etag when uploading the part
```
$ aws --endpoint=http://127.0.0.1:7480 s3api  upload-part --bucket  test --key test-large --part-number 1 --body ./user-config.jam --upload-id 20220922T124511.426602599Z
{
    "ETag": "\"e1e8d0676ebe133a7c9ce47c75011c1e\""
}

$ aws --endpoint=http://127.0.0.1:7480 s3api  upload-part --bucket  test --key test-large --part-number 2 --body ./user-config.jam --upload-id 20220922T124511.426602599Z
{
    "ETag": "\"e1e8d0676ebe133a7c9ce47c75011c1e\""
}
```

Later when doing:
```
aws --endpoint=http://127.0.0.1:7480 s3api complete-multipart-upload --multipart-upload file://parts.test --bucket  test --key test-large --upload-id 20220922T124511.426602599Z
{
    "Location": "http://127.0.0.1:7480/test/test-large",
    "Bucket": "test",
    "Key": "test-large",
    "ETag": ""
}
```

with `parts.test` being:
```
{
"Parts": [
        {
            "PartNumber": 1,
            "ETag": "\"e1e8d0676ebe133a7c9ce47c75011c1e\""
        },
        {
            "PartNumber": 2,
            "ETag": "\"e1e8d0676ebe133a7c9ce47c75011c1e\""
        }
]
}
```
It seems to work fine. So I guess etags being the same is not a crucial thing here.

File: src/rgw/store/sfs/sqlite/objects/object_definitions.h (hunk 7 above)
Comment (jecluis) [reply src/rgw/store/sfs/sqlite/objects/object_definitions.h:65]: it was forgotten there. but yes, it's what I'm looking into now. I'll drop it from the PR.

**Comment by jecluis** [comment]
> I need more time to fully review this but so far it looks good.
> 
> I've been testing and seems to be working fine.
> 
> One thing: I see etags for parts are the same.

RGW is the one doing the etag calculation for parts, and it's essentially an md5 of the part's contents. The only way I see for that to happen (aside from a bug somewhere in the code, of course) is if the parts all have the same contents.

**Comment by 0xavi0** [comment]
yeah, I saw rgw is the one calculating the `etag` and yes: content was the same... so my fault.

File: src/rgw/store/sfs/sqlite/versioned_object/versioned_object_definitions.h [hunk 8]
```diff
@@ -42,6 +43,7 @@ struct DBOPVersionedObjectInfo {
   ceph::real_time creation_time;
   ObjectState object_state;
   std::string version_id;
+  std::string etag;
```
Comment (0xavi0) [inline src/rgw/store/sfs/sqlite/versioned_object/versioned_object_definitions.h:46]: Could you add this new variable to the unit tests?
It's so we don't loose coverage and test all the members.

Something like the following should be enough:
```diff
diff --git a/src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc b/src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc
index 8851597d6e..be30f9164f 100644
--- a/src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc
+++ b/src/test/rgw/sfs/test_rgw_sfs_sqlite_versioned_objects.cc
@@ -116,6 +116,7 @@ DBOPVersionedObjectInfo createTestVersionedObject(uint id, const std::string & o
   test_versioned_object.creation_time = ceph::real_clock::now();
   test_versioned_object.object_state = rgw::sal::ObjectState::OPEN;
   test_versioned_object.version_id = "test_version_id_" + suffix;
+  test_versioned_object.etag = "test_etag_" + suffix;
   return test_versioned_object;
 }
 
@@ -128,6 +129,7 @@ void compareVersionedObjects(const DBOPVersionedObjectInfo & origin, const DBOPV
   ASSERT_EQ(origin.creation_time, dest.creation_time);
   ASSERT_EQ(origin.object_state, dest.object_state);
   ASSERT_EQ(origin.version_id, dest.version_id);
+  ASSERT_EQ(origin.etag, dest.etag);
 }
 
 TEST_F(TestSFSSQLiteVersionedObjects, CreateAndGet) {
```

**Review by 0xavi0 — COMMENTED** [review]
Aside from the unit test... looks good to me.
Let's try to merge this today if possible.

File: src/rgw/store/sfs/sqlite/versioned_object/versioned_object_definitions.h (hunk 8 above)
Comment (jecluis) [reply src/rgw/store/sfs/sqlite/versioned_object/versioned_object_definitions.h:46]: done

**Comment by jhmarina** [comment]
This PR closes https://github.com/aquarist-labs/s3gw/issues/8

**Pull request merged**